### PR TITLE
feat: bad debt cleanup

### PR DESCRIPTION
**Pull request by toteki**
## Description

Automatically repay bad debt using reserves.

closes: #272 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/keeper/reserves.go [hunk 1]
```diff
@@ -28,31 +28,137 @@ func (k Keeper) GetReserveAmount(ctx sdk.Context, denom string) sdk.Int {
...
 	if err != nil {
 		return err
 	}
 
 	store.Set(reserveKey, bz)
 	return nil
 }
+
+// SweepBadDebts attempts to repay all bad debts in the system, efficiently
+// stopping for any denomination where reserves become inqdequate. It should be
```
Comment (toteki) [inline x/leverage/keeper/reserves.go:51]:
```suggestion
// stopping for any denomination where reserves become inadequate. It should be
```

File: x/leverage/types/keys.go [hunk 2]
```diff
@@ -133,3 +135,55 @@ func CreateExchangeRateKeyNoDenom() []byte {
...
+func CreateBadDebtAddressKeyNoAddress(denom string) []byte {
+	// badDebtAddrPrefix | lengthprefixed(denom)
+	var key []byte
+	key = append(key, KeyPrefixBadDebtAddress...)
+	key = append(key, LengthPrefixDenom(denom)...)
+	return key
+}
+
+// LengthPrefixDenom matches the functionality of address.MustLengthPrefix,
+// but we are using it for denoms here.
```
Comment (toteki) [inline x/leverage/types/keys.go:176]:
```suggestion
// but we are using it for denoms here, and also forbid len=0
```

**Comment by toteki** [comment]
Note the simplification reflected in the ADR:

Original design:
- Attempt repay immediately when `collateral` reaches zero
- Mark remaining bad debt if repayment fails
- Sweep remaining bad debt every `InterestEpoch`

As implemented:
- Mark bad debt when `collateral` reaches zero
- Sweep remaining bad debt every `InterestEpoch`

**Comment by alexanderbez** [comment]
Looks great @toteki! I think we could consider combining the two bad indexes into one, but maybe the current design makes the UX cleaner?

File: x/leverage/keeper/reserves.go [hunk 3]
```diff
@@ -28,31 +28,137 @@ func (k Keeper) GetReserveAmount(ctx sdk.Context, denom string) sdk.Int {
...
+
+	// Outer iterator will iterate over all token denoms with bad debt
+	denomPrefix := types.CreateBadDebtDenomKeyNoDenom()
+	denomIter := sdk.KVStorePrefixIterator(store, denomPrefix)
+	defer denomIter.Close()
+
+	for ; denomIter.Valid(); denomIter.Next() {
+		// denomKey is prefix | denom
+		denomKey := denomIter.Key()
+		denom := string(denomKey[len(denomPrefix):]) // remove prefix
```
Comment (adamewozniak) [inline x/leverage/keeper/reserves.go:64]: With the 0x00 byte at the end : 
```suggestion
		denom := string(denomKey[len(denomPrefix):len(denomKey)-1]]) // remove prefix
```

File: x/leverage/types/keys.go [hunk 4]
```diff
@@ -133,3 +135,55 @@ func CreateExchangeRateKeyNoDenom() []byte {
 	key = append(key, KeyPrefixExchangeRate...)
 	return key
 }
+
+// CreateBadDebtDenomKey returns a KVStore key for tracking a denom with unpaid bad debt
+func CreateBadDebtDenomKey(denom string) []byte {
+	// badDebtDenomPrefix | denom | 0x00
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/keys.go:141]: Question - does the null terminator have an advantage, like improving iterators or preventing certain edge cases?

**Comment by toteki** [comment]
> Looks great @toteki! I think we could consider combining the two bad indexes into one, but maybe the current design makes the UX cleaner?

Keeping it in two indices allows for more efficient iteration in the case of reserve exhaustion: 

When the module account can no longer pay back bad debts of a certain denom, the denom list allows it to jump to the next denom, rather than iterating over the remaining debts in the exhausted denoms just to skip repaying them.

That said, the efficiency gain is only there for the reserve exhaustion edge case - I can simplify it to just the `denom | address` list without any other problems.

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**